Sources/FoundationEssentials/Subprocess/Subprocess+Configuration.swift

```diff
@@ -0,0 +1,615 @@
+                } catch {
+                    guard let posixError: POSIXError = error as? POSIXError else {
+                        killError = error
+                        return
+                    }
+                    // Ignore ESRCH (no such process)
+                    if posixError.code != .ESRCH {
+                        killError = error
+                    }
+                }
+            }
+
+            if let inputError = inputError {
+                throw inputError
+            }
+
+            if let outputError = outputError {
+                throw outputError
+            }
+
+            if let errorError = errorError {
+                throw errorError
+            }
+
+            if let killError = killError {
+                throw killError
+            }
+        }
+
+        /// Close each input individually, and throw the first error if there's multiple errors thrown
+        @Sendable
+        internal func cleanupAll(
+            input: ExecutionInput,
+            output: ExecutionOutput,
+            error: ExecutionOutput
+        ) throws {
+            var inputError: Error?
+            var outputError: Error?
+            var errorError: Error?
+
+            do {
+                try input.closeAll()
+            } catch {
+                inputError = error
+            }
+
+            do {
+                try output.closeAll()
+            } catch {
+                outputError = error
+            }
+
+            do {
+                try error.closeAll()
+            } catch {
+                errorError = error
+            }
+
+            if let inputError = inputError {
+                throw inputError
+            }
+            if let outputError = outputError {
+                throw outputError
+            }
+            if let errorError = errorError {
+                throw errorError
+            }
+        }
+
+        internal func run<R>(
+            output: RedirectedOutputMethod,
+            error: RedirectedOutputMethod,
+            _ body: @Sendable @escaping (Subprocess, StandardInputWriter) async throws -> R
+        ) async throws -> Result<R> {
+            let (readFd, writeFd) = try FileDescriptor.pipe()
+            let executionInput: ExecutionInput = .customWrite(readFd, writeFd)
+            let executionOutput: ExecutionOutput = try output.createExecutionOutput()
+            let executionError: ExecutionOutput = try error.createExecutionOutput()
+            let process: Subprocess = try self.spawn(
+                withInput: executionInput,
+                output: executionOutput,
+                error: executionError)
+            // After spawn, cleanup child side fds
+            try self.cleanup(
+                process: process,
+                childSide: true,
+                parentSide: false,
+                attemptToTerminateSubProcess: false
+            )
+            return try await withTaskCancellationHandler {
```

Comment:
IMO this is not the right way to do task cancellation here. I personally think we should check for cancellation only when we do a `read` or `write` syscall. This makes sure that user code in the `body` still continues to execute and only when they try to make a sys call we check for it.